src/repository/opamRepositoryBackend.ml

```diff
@@ -73,25 +73,119 @@ let job_text name label =
+  let file = Filename.concat (OpamFilename.Dir.to_string parent_dir) file in
+  OpamSystem.read file
+
+let lstat parent_dir file =
+  let file = Filename.concat (OpamFilename.Dir.to_string parent_dir) file in
+  Unix.lstat file
+
 let get_diff parent_dir dir1 dir2 =
+  let chrono = OpamConsole.timer () in
   log "diff: %a/{%a,%a}"
     (slog OpamFilename.Dir.to_string) parent_dir
     (slog OpamFilename.Base.to_string) dir1
     (slog OpamFilename.Base.to_string) dir2;
-  let patch = OpamSystem.temp_file ~auto_clean: false "patch" in
-  let patch_file = OpamFilename.of_string patch in
-  let finalise () = OpamFilename.remove patch_file in
-  OpamProcess.Job.catch (fun e -> finalise (); raise e) @@ fun () ->
-  OpamSystem.make_command
-    ~verbose:OpamCoreConfig.(!r.verbose_level >= 2)
-    ~dir:(OpamFilename.Dir.to_string parent_dir) ~stdout:patch
-    "diff"
-    [ "-ruaN";
-      OpamFilename.Base.to_string dir1;
-      OpamFilename.Base.to_string dir2; ]
-  @@> function
-  | { OpamProcess.r_code = 0; _ } -> finalise(); Done None
-  | { OpamProcess.r_code = 1; _ } as r ->
-    OpamProcess.cleanup ~force:true r;
+  let rec aux diffs dir1 dir2 =
+    let files = get_files_for_diff parent_dir dir1 dir2 in
+    let diffs =
+      List.fold_left (fun diffs operation ->
+          let file1, file2 = match operation with
+            | Patch.Delete filename -> (Some filename, None)
+            | Patch.Create filename -> (None, Some filename)
+            | Patch.Edit (file1, file2)
+            | Patch.Rename_only (file1, file2) -> (Some file1, Some file2)
+          in
+          let add_to_diffs content1 content2 diffs =
+            match Patch.diff operation content1 content2 with
+            | None -> diffs
+            | Some diff -> diff :: diffs
+          in
+          match
+            OpamStd.Option.map (lstat parent_dir) file1,
+            OpamStd.Option.map (lstat parent_dir) file2
+          with
+          | Some {Unix.st_kind = Unix.S_REG; _}, None
+          | None, Some {Unix.st_kind = Unix.S_REG; _}
+          | Some {st_kind = S_REG; _}, Some {st_kind = S_REG; _} ->
+            let content1 = Option.map (readfile parent_dir) file1 in
+            let content2 = Option.map (readfile parent_dir) file2 in
+            add_to_diffs content1 content2 diffs
+          | Some {Unix.st_kind = Unix.S_DIR; _}, None
+          | None, Some {Unix.st_kind = Unix.S_DIR; _}
+          | Some {st_kind = S_DIR; _}, Some {st_kind = S_DIR; _} ->
+            aux diffs file1 file2
+          | Some {st_kind = S_DIR; _}, Some {st_kind = S_REG; _} ->
+            let content2 = Option.map (readfile parent_dir) file2 in
+            let diffs = add_to_diffs None content2 diffs in
+            aux diffs file1 None
+          | Some {st_kind = S_REG; _}, Some {st_kind = S_DIR; _} ->
+            let diffs = aux diffs None file2 in
+            let content1 = Option.map (readfile parent_dir) file1 in
+            add_to_diffs content1 None diffs
```

Comment:
Actually these cases are not supported by the unified diff format and thus were never supported before and were just ignored:
```suggestion
          | Some {st_kind = S_DIR; _}, Some {st_kind = S_REG; _} ->
            failwith "Changes from a directory to a regular file are not supported"
          | Some {st_kind = S_REG; _}, Some {st_kind = S_DIR; _} ->
            failwith "Changes from a regular file to a directory are not supported"
```